src/dex/fluid-dex/fluid-dex.ts

```diff
@@ -0,0 +1,881 @@
+import { FluidDexCommonAddresses } from './fluid-dex-generate-pool';
+import { applyTransferFee } from '../../lib/token-transfer-fee';
+import { getDexKeysWithNetwork, getBigIntPow } from '../../utils';
+import { extractReturnAmountPosition } from '../../executor/utils';
+import ResolverABI from '../../abi/fluid-dex/resolver.abi.json';
+import { MultiResult, MultiCallParams } from '../../lib/multi-wrapper';
+import { BytesLike } from 'ethers/lib/utils';
+import {
+  generalDecoder,
+  extractSuccessAndValue,
+  addressDecode,
+} from '../../lib/decoders';
+// @ts-ignore
+import { Tokens } from '../../../tests/constants-e2e';
+
+export class FluidDex extends SimpleExchange implements IDex<FluidDexData> {
+  readonly eventPools: { [id: string]: FluidDexEventPool } = {};
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = false;
+  readonly isFeeOnTransferSupported = false;
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(FluidDexConfig);
+
+  logger: Logger;
+
+  pools: FluidDexPool[];
+
+  readonly FluidCommonAddresses: FluidDexCommonAddresses;
+
+  readonly iFluidDexPool: Interface;
+
+  protected adapters;
+
+  FEE_100_PERCENT = BigInt(1000000);
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.FluidCommonAddresses = new FluidDexCommonAddresses(
+      'FluidDex',
+      FluidDexConfig['FluidDex'][network].commonAddresses,
+      network,
+      dexHelper,
+      this.logger,
+    );
+    this.pools = FluidDexConfig['FluidDex'][network].pools;
+    this.adapters = Adapters[network] || {};
+    this.iFluidDexPool = new Interface(FluidDexPoolABI);
+  }
+
+  async fetchFluidDexPools(blockNumber: number): Promise<FluidDexPool[]> {
+    const poolsFromResolver =
+      await this.FluidCommonAddresses.getStateOrGenerate(blockNumber, false);
+
+    return poolsFromResolver.map(pool => ({
+      id: `FluidDex_${pool.address}`,
+      address: pool.address,
+      token0: pool.token0,
+      token1: pool.token1,
+    }));
+  }
+
+  async updatePoolAndEventPool(blockNumber: number) {
+    this.pools = await this.fetchFluidDexPools(blockNumber);
+    for (const pool of this.pools) {
+      if (!this.eventPools[pool.id]) {
+        this.eventPools[pool.id] = new FluidDexEventPool(
+          'FluidDex',
+          pool.address,
+          this.FluidCommonAddresses.commonAddresses,
+          this.network,
+          this.dexHelper,
+          this.logger,
+        );
+      }
+    }
+  }
+
+  // Initialize pricing is called once in the start of
+  // pricing service. It is intended to setup the integration
+  // for pricing requests. It is optional for a DEX to
+  // implement this function
+  async initializePricing(blockNumber: number) {
+    Object.entries(this.eventPools).forEach(([id, eventPool]) => {
```

Comment:
there are no entries of this.eventPools at this time right? initializePricing is called at the very beginning, how can there be any eventPools? 

Also why is eventPools readonly if we are writing to it in updatePoolAndEventPool?

Reply:
changed 👍